src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Block.cs

```diff
@@ -284,15 +301,31 @@ private bool CData()
             }
             else
             {
-                var tagName = String.Empty;
-                if (At(HtmlSymbolType.Text))
+                var tagName = string.Empty;
+                HtmlSymbol bangSymbol = null;
+
+                if (At(HtmlSymbolType.Bang))
+                {
+                    bangSymbol = CurrentSymbol;
+
+                    var nextSymbol = Lookahead();
+
+                    if (nextSymbol.Type == HtmlSymbolType.Text)
+                    {
+                        tagName = nextSymbol.Content;
+                    }
+                }
+                else if (At(HtmlSymbolType.Text))
```

Comment:
It doesn't throw it away, this entire section is only looking ahead. the section below is what captures the `!` tag (really anything after the `/`) by calling into `AcceptUntil(HtmlSymbolType.CloseAngle)`.